base_layer/core/src/base_node/sync/header_sync/synchronizer.rs

```diff
@@ -0,0 +1,429 @@
+                return Err(err.into());
+            },
+            Err(err) => {
+                return Err(err.into());
+            },
+        };
+        let proto::FindChainSplitResponse {
+            headers,
+            fork_hash_index,
+        } = resp;
+        debug!(
+            target: LOG_TARGET,
+            "Found split {} blocks back, received {} headers from peer `{}`",
+            fork_hash_index,
+            headers.len(),
+            peer
+        );
+
+        if fork_hash_index >= block_hashes.len() as u32 {
+            let _ = self
+                .ban_peer(peer.clone(), BanReason::SplitHashGreaterThanHashes {
+                    fork_hash_index,
+                    num_block_hashes: block_hashes.len(),
+                })
+                .await;
+            return Err(BlockHeaderSyncError::FoundHashIndexOutOfRange(
+                block_hashes.len() as u32,
+                fork_hash_index,
+            ));
+        }
+
+        // If the peer returned no new headers, this means header sync is done.
+        if headers.is_empty() {
+            if fork_hash_index > 0 {
+                debug!(
+                    target: LOG_TARGET,
+                    "Peer `{}` has sent no headers but forked_hash_index is {}. The peer is likely behind our chain.",
+                    peer,
+                    fork_hash_index
+                );
+                // Another peer will be attempted, if possible
+                return Err(BlockHeaderSyncError::NotInSync);
+            }
+
+            debug!(target: LOG_TARGET, "Already in sync with peer `{}`.", peer);
+            return Ok(false);
+        }
+
+        if headers.len() > NUM_HEADERS_TO_REQUEST as usize {
+            self.ban_peer(peer.clone(), BanReason::PeerSentTooManyHeaders(headers.len()))
+                .await?;
+            return Err(BlockHeaderSyncError::NotInSync);
+        }
+
+        // Overflow has already been checked above
+        let fork_hash_index = fork_hash_index as usize;
+        // We can trust that the header associated with this hash exists because block_hashes is data this node supplied
+        let chain_split_hash = block_hashes[fork_hash_index].clone();
+
+        let headers = headers
+            .into_iter()
+            .map(BlockHeader::try_from)
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(BlockHeaderSyncError::ReceivedInvalidHeader)?;
+        let num_new_headers = headers.len();
+
+        self.header_validator.initialize_state(chain_split_hash).await?;
+        for header in &headers {
+            debug!(
+                target: LOG_TARGET,
+                "Validating header #{} (Pow: {})",
+                header.height,
+                header.pow_algo(),
+            );
+            self.header_validator.validate(header)?;
+            debug!(target: LOG_TARGET, "Header #{} is VALID", header.height,)
+        }
+
+        self.header_validator
```

Comment:
This check is not sufficient. In your example it checks `I2` against `I1` however this is not sufficient to know if the PoW is stronger between the two chains - the only way to know this AFAIK is to download all headers and compare the strength of each fork. I think I put a TODO comment just after this line about this. Header download and validation can be very quick (few seconds) so it may be worth asking the peer for their chain metadata, then if they say they have stronger PoW, downloading headers and checking.

Reply:
That is basically what is happening now, I think the only way to know is to download all of the headers that the peer has before doing this check

Reply:
This check is here as an attempt to ensure that the peer can provide some good headers before we make any changes to our local state (rewinding the headers). We might need a "staging" area in the database for headers before they are accepted in so that tips can be compared.

Side note: header validation to ~6000 takes 5 minutes due to the current implementation of randomX validation - if that can be optimised ~6000 header download + validation would take around 30seconds.